risc0's fork of crypto-bigint does not build with `serde` when target is zkVM

Can't file an issue at https://github.com/risc0/RustCrypto-crypto-bigint so here I am.
I'm using Rust stable (1.73). Easiest repro is to install the `risc0` Rust toolchain, then build that crate with `serde` targeting `zkvm`:

```sh
$ cd RustCrypto-crypto-bigint
$ cargo +risc0 build --target riscv32im-risc0-zkvm-elf --features serde
```
```
   Compiling serde v1.0.164
   Compiling cfg-if v1.0.0
   Compiling base16ct v0.2.0
   Compiling subtle v2.5.0
   Compiling getrandom v0.2.10
   Compiling rand_core v0.6.4
   Compiling serdect v0.2.0
   Compiling crypto-bigint v0.5.2 (/Users/moz/workspace/risc0-rustcrypto-crypto-bigint)
error[E0401]: can't use generic parameters from outer function
   --> src/uint/modular/constant_mod.rs:226:39
    |
213 | impl<'de, MOD, const LIMBS: usize> Deserialize<'de> for Residue<MOD, LIMBS>
    |                      ----- const parameter from outer function
...
226 |                     const R_INV: Uint<LIMBS> = &MOD::R.inv_odd_mod(&MOD::MODULUS);
    |                                       ^^^^^ use of generic parameter from outer function

error[E0401]: can't use generic parameters from outer function
   --> src/uint/modular/constant_mod.rs:226:49
    |
213 | impl<'de, MOD, const LIMBS: usize> Deserialize<'de> for Residue<MOD, LIMBS>
    |           --- type parameter from outer function
...
226 |                     const R_INV: Uint<LIMBS> = &MOD::R.inv_odd_mod(&MOD::MODULUS);
    |                                                 ^^^^^^ use of generic parameter from outer function

error[E0401]: can't use generic parameters from outer function
   --> src/uint/modular/constant_mod.rs:226:69
    |
213 | impl<'de, MOD, const LIMBS: usize> Deserialize<'de> for Residue<MOD, LIMBS>
    |           --- type parameter from outer function
...
226 |                     const R_INV: Uint<LIMBS> = &MOD::R.inv_odd_mod(&MOD::MODULUS);
    |                                                                     ^^^^^^^^^^^^ use of generic parameter from outer function

error[E0747]: unresolved item provided when a constant was expected
   --> src/uint/modular/constant_mod.rs:226:39
    |
226 |                     const R_INV: Uint<LIMBS> = &MOD::R.inv_odd_mod(&MOD::MODULUS);
    |                                       ^^^^^
    |
help: if this generic argument was intended as a const parameter, surround it with braces
    |
226 |                     const R_INV: Uint<{ LIMBS }> = &MOD::R.inv_odd_mod(&MOD::MODULUS);
    |                                       +       +

Some errors have detailed explanations: E0401, E0747.
For more information about an error, try `rustc --explain E0401`.
error: could not compile `crypto-bigint` (lib) due to 4 previous errors
```

--- a/.github/workflows/crypto-bigint.yml
+++ b/.github/workflows/crypto-bigint.yml
@@ -1,5 +1,9 @@
 name: crypto-bigint
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
+
 on:
   pull_request:
     paths-ignore:
@@ -27,7 +31,7 @@ jobs:
           - wasm32-unknown-unknown
     steps:
       - uses: actions/checkout@v3
-      - uses: dtolnay/rust-toolchain@master
+      - uses: risc0/risc0/.github/actions/rustup@v0.19.1
         with:
           toolchain: ${{ matrix.rust }}
           targets: ${{ matrix.target }}
@@ -61,7 +65,7 @@ jobs:
             rust: stable
     steps:
       - uses: actions/checkout@v3
-      - uses: dtolnay/rust-toolchain@master
+      - uses: risc0/risc0/.github/actions/rustup@v0.19.1
         with:
           toolchain: ${{ matrix.rust }}
           targets: ${{ matrix.target }}
@@ -87,7 +91,7 @@ jobs:
     steps:
       - uses: actions/checkout@v3
       - run: ${{ matrix.deps }}
-      - uses: dtolnay/rust-toolchain@master
+      - uses: risc0/risc0/.github/actions/rustup@v0.19.1
         with:
           toolchain: ${{ matrix.rust }}
           targets: ${{ matrix.target }}
@@ -101,7 +105,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
-      - uses: dtolnay/rust-toolchain@master
+      - uses: risc0/risc0/.github/actions/rustup@v0.19.1
         with:
           toolchain: stable
       - run: cargo run --release
@@ -113,7 +117,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
-      - uses: dtolnay/rust-toolchain@master
+      - uses: risc0/risc0/.github/actions/rustup@v0.19.1
         with:
           toolchain: 1.65.0
           components: clippy
@@ -123,7 +127,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
-      - uses: dtolnay/rust-toolchain@master
+      - uses: risc0/risc0/.github/actions/rustup@v0.19.1
         with:
           toolchain: stable
           components: rustfmt
@@ -133,7 +137,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
-      - uses: dtolnay/rust-toolchain@master
+      - uses: risc0/risc0/.github/actions/rustup@v0.19.1
         with:
           toolchain: 1.65.0
       - run: cargo build --all-features --benches
@@ -142,7 +146,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
-      - uses: dtolnay/rust-toolchain@master
+      - uses: risc0/risc0/.github/actions/rustup@v0.19.1
         with:
           toolchain: stable
       - run: cargo doc --all-features

--- a/src/uint/modular/constant_mod.rs
+++ b/src/uint/modular/constant_mod.rs
@@ -219,17 +219,18 @@ where
     where
         D: Deserializer<'de>,
     {
+        #[cfg(all(target_os = "zkvm", target_arch = "riscv32"))]
+        let r_inv: Uint<{ LIMBS }> = MOD::R.inv_odd_mod(&MOD::MODULUS).0;
+
         Uint::<LIMBS>::deserialize(deserializer).and_then(|montgomery_form| {
             if Uint::ct_lt(&montgomery_form, &MOD::MODULUS).into() {
                 #[cfg(all(target_os = "zkvm", target_arch = "riscv32"))]
                 if LIMBS == risc0::BIGINT_WIDTH_WORDS {
-                    const R_INV: Uint<LIMBS> = &MOD::R.inv_odd_mod(&MOD::MODULUS);
-
                     // In the RISC Zero zkVM 256-bit residues are represented in standard form.
                     // To ensure this is interoperable with the host, convert to standard form.
-                    let value = risc0::modmul_uint_256(&montgomery_form, &R_INV, &MOD::MODULUS);
+                    let value = risc0::modmul_uint_256(&montgomery_form, &r_inv, &MOD::MODULUS);
                     return Ok(Self {
-                        value,
+                        montgomery_form: value,
                         phantom: PhantomData,
                     });
                 }
@@ -264,5 +265,27 @@ where
         }
 
         self.montgomery_form.serialize(serializer)
+    }
+}
+
+#[cfg(all(test, feature = "serde"))]
+mod tests {
+    use crate::{const_residue, impl_modulus, modular::constant_mod::ResidueParams, U256};
+
+    impl_modulus!(
+        Modulus,
+        U256,
+        "9CC24C5DF431A864188AB905AC751B727C9447A8E99E6366E1AD78A21E8D882B"
+    );
+
+    #[test]
+    fn serde_roundtrip() {
+        let value_uint = U256::from(105u64);
+        let value = const_residue!(value_uint, Modulus);
+
+        let value_encoded = bincode::serialize(&value).unwrap();
+        let value_decoded = bincode::deserialize(&value_encoded).unwrap();
+
+        assert_eq!(value, value_decoded);
     }
 }

--- a/src/uint/modular/inv.rs
+++ b/src/uint/modular/inv.rs
@@ -17,7 +17,7 @@ pub fn inv_montgomery_form<const LIMBS: usize>(
     }
 
     (
-        mul_montgomery_form(&inverse, &r3, modulus, mod_neg_inv),
+        mul_montgomery_form(&inverse, r3, modulus, mod_neg_inv),
         is_some,
     )
 }

